File: Frameworks/react-web/tests/unit/models/ods-help.test.ts
import { describe, it, expect } from 'vitest'
import { parseHelp, parseTourStep } from '../../../src/models/ods-help.ts'

// ===========================================================================
// OdsHelp model tests
// ===========================================================================

describe('parseHelp', () => {
  // -------------------------------------------------------------------------
  // Null / missing / invalid input
  // -------------------------------------------------------------------------

  it('returns undefined for null', () => {
    expect(parseHelp(null)).toBeUndefined()
  })

  it('returns undefined for undefined', () => {
    expect(parseHelp(undefined)).toBeUndefined()
  })

  it('returns undefined for non-object', () => {
    expect(parseHelp('not-an-object')).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // Valid help objects
  // -------------------------------------------------------------------------

  it('parses valid help with overview and pages', () => {
    const help = parseHelp({
      overview: 'Welcome to the app.',
      pages: { home: 'Home page help', settings: 'Settings help' },
    })
    expect(help).toBeDefined()
    expect(help!.overview).toBe('Welcome to the app.')
    expect(help!.pages).toEqual({ home: 'Home page help', settings: 'Settings help' })
  })

  it('defaults pages to empty object when missing', () => {
    const help = parseHelp({ overview: 'Some overview' })
    expect(help).toBeDefined()
    expect(help!.pages).toEqual({})
  })
})

// ===========================================================================
// OdsTourStep model tests
// ===========================================================================

describe('parseTourStep', () => {
  it('parses all fields', () => {
    const step = parseTourStep({
      title: 'Step 1',
      content: 'Click the button.',
      page: 'home',
    })
    expect(step.title).toBe('Step 1')
    expect(step.content).toBe('Click the button.')
    expect(step.page).toBe('home')
  })

  it('page is undefined when missing', () => {
    const step = parseTourStep({
      title: 'Step 2',
      content: 'Enter your name.',
    })
    expect(step.title).toBe('Step 2')
    expect(step.content).toBe('Enter your name.')
    expect(step.page).toBeUndefined()
  })
})

File: Frameworks/flutter-local/test/models/ods_help_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/models/ods_help.dart';

void main() {
  // =========================================================================
  // OdsHelp.fromJson
  // =========================================================================

  group('OdsHelp.fromJson', () {
    test('valid help with overview and pages', () {
      final help = OdsHelp.fromJson({
        'overview': 'Welcome to the app.',
        'pages': {
          'home': 'This is the home page.',
          'settings': 'Configure your preferences here.',
        },
      });
      expect(help.overview, 'Welcome to the app.');
      expect(help.pages, hasLength(2));
      expect(help.pages['home'], 'This is the home page.');
      expect(help.pages['settings'], 'Configure your preferences here.');
    });

    test('missing pages defaults to empty map', () {
      final help = OdsHelp.fromJson({
        'overview': 'An app with no page help.',
      });
      expect(help.pages, isEmpty);
    });
  });

  // =========================================================================
  // OdsTourStep.fromJson
  // =========================================================================

  group('OdsTourStep.fromJson', () {
    test('all fields including page', () {
      final step = OdsTourStep.fromJson({
        'title': 'Getting Started',
        'content': 'Here is how you begin.',
        'page': 'page_intro',
      });
      expect(step.title, 'Getting Started');
      expect(step.content, 'Here is how you begin.');
      expect(step.page, 'page_intro');
    });

    test('missing page is null', () {
      final step = OdsTourStep.fromJson({
        'title': 'Final Notes',
        'content': 'That is everything you need to know.',
      });
      expect(step.title, 'Final Notes');
      expect(step.content, 'That is everything you need to know.');
      expect(step.page, isNull);
    });
  });
}
